Hitomi: Add Image Format Preferences

KenjieDec opened the pull request:
Closes #4030
Checklist:
- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [ ] Have removed `web_hi_res_512.png` when adding a new extension

KenjieDec committed:
Add image filetype preferences

KenjieDec committed:
Update extVersionCode

AwkwardPeak7 commented on src/all/hitomi/src/eu/kanade/tachiyomi/extension/all/hitomi/Hitomi.kt:
```diff
@@ -567,14 +590,24 @@ class Hitomi(
             val hash = img.hash
+
+            val avif = img.hasavif == 1 && imageType() == "avif"
+            val jxl = img.hasjxl == 1 && imageType() == "jxl"
```
```suggestion
            val typePref = imageType()
            val avif = img.hasavif == 1 && typePref == "avif"
            val jxl = img.hasjxl == 1 && itypePref == "jxl"
```

AwkwardPeak7 commented on src/all/hitomi/src/eu/kanade/tachiyomi/extension/all/hitomi/Hitomi.kt:
```diff
@@ -657,11 +690,40 @@ class Hitomi(
+            .body(body)
+            .header("Content-Type", type)
+            .build()
+    }
```
consider checking the file extension as well if it is equal to jxl

AwkwardPeak7 commented on src/all/hitomi/src/eu/kanade/tachiyomi/extension/all/hitomi/Hitomi.kt:
```diff
@@ -507,7 +528,9 @@ class Hitomi(
-            language?.let { append("Language: ", language) }
+            language?.let { append("Language: ", language, "\n") }
+
+            append("Available Image Types: ", listOfNotNull(webp, avif, jxl).joinToString())
```
not really needed?

AwkwardPeak7 commented on src/all/hitomi/src/eu/kanade/tachiyomi/extension/all/hitomi/Hitomi.kt:
```diff
@@ -657,11 +690,40 @@ class Hitomi(
+            title = "Images Type"
+            entries = arrayOf("webp", "avif", "jxl")
+            entryValues = arrayOf("webp", "avif", "jxl")
+            summary = "%s"
```
"Clear chapter cache to apply changes"

KenjieDec committed:
Apply suggestions

- Confirm jxl by using its signatures when passing through interceptor
- & others

KenjieDec committed:
Update HitomiDto.kt

KenjieDec committed:
Fix

AwkwardPeak7 commented on src/all/hitomi/src/eu/kanade/tachiyomi/extension/all/hitomi/Hitomi.kt:
```diff
@@ -657,11 +685,53 @@ class Hitomi(
+            return response
+        }
+
+        val bytesArray = response.body.bytes()
```
this will consume the response body so returning it as response later cause exception

KenjieDec replied:
I don't understand, can you please explain more? Thank you.
Sorry if this is a weird question 🙏

AwkwardPeak7 replied:
calling `response.body.*` methods like `.string()` or `.bytes()` etc, "consume" the body content and close the reponse, so it no longer will have the body contents. 
https://square.github.io/okhttp/5.x/okhttp/okhttp3/-response/index.html#1708455149%2FProperties%2F1481842020

If you want to check the response without closing it, [peek into it.](https://github.com/keiyoushi/extensions-source/blob/d7dc91fd8b9b593c1e49a5a4a832991b8813f79a/src/en/luascans/src/eu/kanade/tachiyomi/extension/en/luascans/LuaScans.kt#L26)

KenjieDec replied:
Thanks for explaining it to me :pray: 
Sorry, but I think the commit will be pushed a bit late because I don't have a pc with me currently.

KenjieDec committed:
Fix - Apply Suggestion

AwkwardPeak7 commented on src/all/hitomi/src/eu/kanade/tachiyomi/extension/all/hitomi/Hitomi.kt:
```diff
@@ -657,11 +685,53 @@ class Hitomi(
+            return response
+        }
+
+        val bytesArray = response.peekBody(Long.MAX_VALUE).bytes()
```
while this works you could optimize it to only peek the necessary bytes

vetleledaal replied:
A possible implementation of this would be:
```suggestion
        val bytesPeek = max(signatureOne.size, signatureTwo.size).toLong()
        val bytesArray = response.peekBody(bytesPeek).bytes()
```

KenjieDec committed:
Apply suggestion

Co-authored-by: Vetle Ledaal <vetle.ledaal@gmail.com>

KenjieDec committed:
Fix

KenjieDec committed:
Lint Fix?
